<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Breakmer by a-bioinformatician</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Breakmer</h1>
        <p>A method to identify structural variation from sequencing data in target regions</p>

        <p class="view"><a href="https://github.com/a-bioinformatician/BreaKmer">View the Project on GitHub <small>a-bioinformatician/BreaKmer</small></a></p>


        <ul>
          <li><a href="https://github.com/a-bioinformatician/BreaKmer/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/a-bioinformatician/BreaKmer/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/a-bioinformatician/BreaKmer">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>
<a id="breakmer" class="anchor" href="#breakmer" aria-hidden="true"><span class="octicon octicon-link"></span></a>BreaKmer</h1>

<p>A method to identify genomic structural variation in target regions/genes from reference-aligned high-throughput sequence data. It uses a “kmer” strategy to assemble misaligned sequence reads for predicting insertions, deletions, inversions, tandem duplications, and translocations at base-pair resolution.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>The following are required for installation:</p>

<ul>
<li><a href="https://www.python.org/download/releases/2.7">Python 2.7</a></li>
</ul>

<p>Download the python scripts and run the command:</p>

<pre><code>python setup.py install
</code></pre>

<p>Use appropriate commands for installing locally:</p>

<pre><code>python setup.py install --user
</code></pre>

<p>Using the setup.py script for installation should setup the required python module dependencies for appropriate usage. If the install script is not used, the two modules need to be downloaded and installed if not already:</p>

<ul>
<li><a href="http://biopython.org/wiki/Main_Page">Biopython 1.62</a></li>
<li><a href="https://code.google.com/p/pysam/">Pysam 0.6</a></li>
</ul>

<p>BreaKmer has currently been installed and tested on:</p>

<ul>
<li>64bit linux using CentOS release 5.5 and python2.7.2 </li>
<li>64bit linux using Ubuntu release 14.04 and python2.7.6</li>
</ul>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<p>List the available command line parameters.</p>

<pre><code>python &lt;PATH_TO_BREAKMER_DIR&gt;/breakmer.py -h
</code></pre>

<p>Analyze all the target genes specified in the targets bed file.</p>

<pre><code>python &lt;path to breakmer scripts&gt;/breakmer.py run &lt;options&gt; -c &lt;path to breakmer configuration file&gt;
</code></pre>

<p>Prepare the reference data files before starting the analysis.</p>

<pre><code>python &lt;path to breakmer scripts&gt;/breakmer.py prepare_reference_data &lt;options&gt; -c &lt;path to breakmer configuration file&gt;
</code></pre>

<p>Start the blat server on a specific host and port.</p>

<pre><code>python &lt;path to breakmer scripts&gt;/breakmer.py start_blat_server &lt;options&gt; -c &lt;path to breakmer configuration file&gt;
</code></pre>

<p>Analyze a subset of genes specified in a file.</p>

<pre><code>python &lt;path to breakmer scripts&gt;/breakmer.py -g &lt;file containing list of target genes to analyze&gt; &lt;path to breakmer configuration file&gt;
</code></pre>

<h2>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span class="octicon octicon-link"></span></a>Requirements</h2>

<h3>
<a id="programs" class="anchor" href="#programs" aria-hidden="true"><span class="octicon octicon-link"></span></a>Programs</h3>

<ul>
<li>BLAT standalone and server binaries (<a href="http://hgdownload.cse.ucsc.edu/admin/exe/">blat, gfServer, gfClient, faToTwoBit</a>).

<ul>
<li>Re-alignment to reference sequence.</li>
<li>Versions tested :

<ul>
<li>standalone BLAT v35x1</li>
<li>gfServer v35x1</li>
<li>gfClient v35x1</li>
</ul>
</li>
</ul>
</li>
<li>
<a href="https://code.google.com/p/cutadapt/">Cutadapt</a>

<ul>
<li>Trims adapter sequence from aligned reads.</li>
<li>v1.5 tested</li>
</ul>
</li>
<li>
<a href="http://www.cbcb.umd.edu/software/jellyfish">Jellyfish</a>

<ul>
<li>Generating kmers.</li>
<li>
<a href="http://www.cbcb.umd.edu/software/jellyfish/jellyfish-1.1.11.tar.gz">v1.1.11</a> tested</li>
<li>
<a href="http://www.genome.umd.edu/jellyfish.html">v2.1.3</a> tested</li>
</ul>
</li>
</ul>

<p>When these programs are installed, the paths to the binaries can be either be specified in the BreaKmer configuration file or put in the path (e.g. export PATH=$PATH:/path/to/binary).</p>

<h3>
<a id="sequence-data" class="anchor" href="#sequence-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sequence data</h3>

<ul>
<li>Sample bam file

<ul>
<li>BreaKmer requires sequence reads that have been aligned to a reference sequence in a binary alignment format (BAM). The alignment program must soft-clip or trim reads that can be partially aligned to the reference sequence (e.g., bwa, bowtie, novoalign, mosaik). The partially aligned reads and unmapped reads with mapped mates (paired-end data) are used to build contigs with potential SV. This bam file is required as input in the configuration file as the "sample_bam_file".

<ul>
<li>The BAM files need to be sorted and indexed, with the indexed files in the same directory as the BAM file.</li>
<li>There are a number of aligners that soft-clip partially aligned sequences, bwa and Bowtie are two well-known tools:

<ul>
<li><a href="http://bio-bwa.sourceforge.net/">bwa</a></li>
<li><a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a></li>
</ul>
</li>
<li>It is also useful to mark duplicate reads prior to using BreaKmer, as these will inflate read counts for identified variants.

<ul>
<li><a href="http://picard.sourceforge.net/command-line-overview.shtml#MarkDuplicates">Picard MarkDuplicates</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li>Normal bam file

<ul>
<li>A matched normal with similarly targeted sequencing data as the sample can be input to help filter germline events. The sequences in the normal bam file are processed for each target in a similar manner to the sample sequences and used to further filter our kmers. This is an optional input in the configuration file as "normal_bam_file".</li>
</ul>
</li>
</ul>

<h3>
<a id="reference-data" class="anchor" href="#reference-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Reference data</h3>

<ul>
<li>Reference fasta file 

<ul>
<li>BreaKmer makes use of reference sequence data throughout the program. The genomic reference sequence used to align the short sequence reads is required as an input in the configuration file as the "reference_fasta".</li>
<li>Format: single fasta file with chromosome number/id as names (i.e. '&gt;1', '&gt;2', '&gt;3')</li>
<li>Hg19 fasta files can be downloaded from UCSC Genome Browser(<a href="http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/">http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/</a>)</li>
<li>This file should be placed in a writeable directory. A 2bit file will be generated from this file to start the blat server.</li>
</ul>
</li>
<li>Reference genome annotation 

<ul>
<li>This file is used by BreaKmer to annotate the locations of the breakpoints identified and the genes involved. This is required as input in the configuration file as the "gene_annotation_file" only when annotation is desired for the output files (currently not implemented) or when the --generate_image option is set and the corresponding genes and transcripts of the variant are to be visualized.</li>
<li>Format: GTF formatted file (see below).</li>
<li>Gencode transcript annotation file can be downloaded from <a href="http://www.gencodegenes.org/releases/19.html">Gencode Release 19 (GRCh37.p13)</a>.</li>
</ul>
</li>
</ul>

<h3>
<a id="other-files" class="anchor" href="#other-files" aria-hidden="true"><span class="octicon octicon-link"></span></a>Other files</h3>

<ul>
<li>Target regions bed file

<ul>
<li>The target regions sequenced need to be specified in a bed file for BreaKmer to know the coordinates of the regions sequenced and which are to be analyzed. This is a required input in the configuration file as the "targets_bed_file".</li>
<li>Format: a tab delimited file containing the chromosome, start, end, HUGO gene name, and exon/intron feature of a tile region of the genome. The tiled regions should not be overlapping.</li>
</ul>
</li>
</ul>

<h2>
<a id="configuration-file" class="anchor" href="#configuration-file" aria-hidden="true"><span class="octicon octicon-link"></span></a>Configuration file</h2>

<ul>
<li>A template configuration file with all the options is available, breakmer.cfg.</li>
<li>Below lists and describes the parameters that can be used in the configuration file. Do not keep commented text (i.e., #Required parameters) in the configuration file.</li>
<li>Note that the paths to the six required program binaries (Cutadapt, Jellyfish, blat, gfServer, and gfClient) can be set in the configuration file
or these binaries can be included in the users path (e.g., for linux users: export PATH=$PATH:/path/to/binary).</li>
<li>Use full paths (e.g., /home/bob and not ~/)</li>
</ul>

<pre><code># Required parameters
analysis_name=&lt;sample_id, string value that all the output files will contain&gt;
targets_bed_file=&lt;path to bed file containing locations of target regions&gt;
sample_bam_file=&lt;path to sorted and indexed sample bam file&gt;
analysis_dir=&lt;path to analysis directory, location where analysis and final output files will be located&gt;
reference_data_dir=&lt;path to where reference files will/are stored for each of the targeted genes analyzed&gt; 
cutadapt=&lt;path to cutadapt binary v1.5, i.e. /usr/bin/cutadapt-1.5/bin/cutadapt&gt; 
cutadapt_config_file=&lt;path to cutadapt configuration file&gt; 
jellyfish=&lt;path to Jellyfish binary, i.e. /usr/bin/jellyfish&gt;
blat=&lt;path to blat binary, i.e. /usr/bin/blat&gt;
gfserver=&lt;path to gfServer binary, i.e. /usr/bin/gfServer&gt;
gfclient=&lt;path to gfClient binary, i.e. /usr/bin/gfClient&gt;
fatotwobit=&lt;path to faToTwoBit binary, i.e. /usr/bin/faToTwoBit&gt;
reference_fasta=&lt;path to whole genome reference fasta file, one file with all records&gt;
gene_annotation_file=&lt;path to gene annotation file&gt;
kmer_size=15

# Optional parameters
other_regions_file=&lt;path to bed file containing coordinates for targeted unannotated cluster regions if they exist, such as IGH, IGK&gt; 
normal_bam_file=&lt;path to normal bam file, can be used to filter germline events with matched-normal sample&gt;
</code></pre>

<h2>
<a id="input-file-formats" class="anchor" href="#input-file-formats" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input file formats</h2>

<ul>
<li>targets_bed_file = tab-delimited file with columns: chr, start, end, region_name, feature_name

<ul>
<li>Each "target" region can contain multiple subregions that are annotated by feature type (i.e., intron, exon). These feature types are used in the filtering steps with certain parameters. Note that the minimum coordinate and maximum coordinate +/- 200 bp are used as the boundaries for each region (e.g., 45003745-200,45008540+200 for B2M).</li>
</ul>
</li>
</ul>

<pre><code>15      45003745        45003811        B2M     exon
15      45007621        45007922        B2M     exon
15      45008527        45008540        B2M     exon
</code></pre>

<ul>
<li>other_regions_file = tab-delimited file with columns: chr, start, end, region_name

<ul>
<li>This file is intended to cover regions that are not annotated in the annotation file. These are useful for cluster regions that are not well annotated in the annotation files.</li>
</ul>
</li>
</ul>

<pre><code>14   22090057        23021075        TRA
7    141998851       142510972       TRB
</code></pre>

<ul>
<li>cutadapt_config_file = each row corresponds to a parameter for cutadapt (see cutadapt.cfg example file or cutadapt documentation)

<ul>
<li>The file provided is intended for data generated using the paired-end Illumina TruSeq library.</li>
<li>Many of the Illumina library sequences have been annotated <a href="https://wikis.utexas.edu/display/GSAF/Illumina+-+all+flavors">elsewhere</a>.</li>
</ul>
</li>
<li>reference_fasta = genome reference fasta formatted file containing all the chromosome reference sequences that were used to initially align the data.

<ul>
<li>This shold be a single file containing all the sequences.</li>
<li>Note that the chromosome IDs should match the chromosome IDs of the alignment file (i.e., using chr1 vs. 1).</li>
</ul>
</li>
<li>gene_annotation = Annotation file containing the genomic coordinates for annotated gene transcripts and their corresponding exon regions. These can be downloaded from Gencode, (i.e., <a href="http://www.gencodegenes.org/releases/19.html">Gencode Release 19 (GRCh37.p13)</a>)</li>
</ul>

<pre><code>column-number content values/format
1 chromosome name chr{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y,M}
2 annotation source {ENSEMBL,HAVANA}
3 feature-type  {gene,transcript,exon,CDS,UTR,start_codon,stop_codon,Selenocysteine}
4 genomic start location  integer-value (1-based)
5 genomic end location  integer-value
6 score (not used)  .
7 genomic strand  {+,-}
8 genomic phase (for CDS features)  {0,1,2,.}
9 additional information as key-value pairs see below

chr21   HAVANA  transcript      10862622        10863067        .       +       .       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
chr21   HAVANA  exon    10862622        10862667        .       +       .       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
chr21   HAVANA  CDS     10862622        10862667        .       +       0       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
chr21   HAVANA  start_codon     10862622        10862624        .       +       0       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
chr21   HAVANA  exon    10862751        10863067        .       +       .       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
chr21   HAVANA  CDS     10862751        10863064        .       +       2       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
chr21   HAVANA  stop_codon      10863065        10863067        .       +       0       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
chr21   HAVANA  UTR     10863065        10863067        .       +       .       gene_id "ENSG00000169861"; transcript_id "ENST00000302092"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "IGHV1OR15-5"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "IGHV1OR15-5-001"; level 2; havana_gene "OTTHUMG00000074130"; havana_transcript "OTTHUMT00000157419";
...
</code></pre>

<h2>
<a id="breakmer-parameters" class="anchor" href="#breakmer-parameters" aria-hidden="true"><span class="octicon octicon-link"></span></a>BreaKmer parameters</h2>

<table>
<thead>
<tr>
<th>Program Mode</th>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td>run</td>
<td></td>
<td>Perform structural variant detection analysis</td>
<td>NA</td>
</tr>
<tr>
<td></td>
<td>-h, --help</td>
<td>Show the help message and exit</td>
<td>NA</td>
</tr>
<tr>
<td></td>
<td>--log_level</td>
<td>The level of logging detail to record while program is running.</td>
<td>DEBUG</td>
</tr>
<tr>
<td></td>
<td>--indel_size</td>
<td>The minimum indel size (in base pairs) to keep and output.</td>
<td>15</td>
</tr>
<tr>
<td></td>
<td>--trl_sr_thresh</td>
<td>Split read support threshold for translocation events (i.e., the minimum number of split reads required to keep a translocation event).</td>
<td>2</td>
</tr>
<tr>
<td></td>
<td>--indel_sr_thresh</td>
<td>Split read support threshold for indels</td>
<td>5</td>
</tr>
<tr>
<td></td>
<td>--rearr_sr_thresh</td>
<td>Split read support threshold for general rearrangements, inversions, tandem duplications, or other unclassified rearrangements.</td>
<td>2</td>
</tr>
<tr>
<td></td>
<td>--rearr_min_seg_len</td>
<td>Minimum length (in base pairs) of a realignment portion of an assembled contig sequence to consider for filtering a rearrangement event (inversions, tandem duplications, other unclassified rearrangements).</td>
<td>30</td>
</tr>
<tr>
<td></td>
<td>--trl_min_seg_len</td>
<td>Minimum length (in base pairs) of a realignment portion of an assembled contig sequence to consider for filtering a translocation event.</td>
<td>25</td>
</tr>
<tr>
<td></td>
<td>--align_thresh</td>
<td>Threshold for minimum read alignment during assembly.</td>
<td>0.9</td>
</tr>
<tr>
<td></td>
<td>--no_output_header</td>
<td>Suppress column headers in output files.</td>
<td>False</td>
</tr>
<tr>
<td></td>
<td>--discread_only_thresh</td>
<td>The number of discordant read pairs in a cluster to output without evidence from a split read event.</td>
<td>2</td>
</tr>
<tr>
<td></td>
<td>--generate_image</td>
<td>Generate a plot containing the assembled reads used to create a contig sequence with a called variant along with genes and transcripts associated with the event - requires gene_annotation_file in configuration file to be set with GTF file containing transcript annotations.</td>
<td>False</td>
</tr>
<tr>
<td></td>
<td>--hostname</td>
<td>The hostname for the blat server.</td>
<td>localhost</td>
</tr>
<tr>
<td></td>
<td>-g, --gene_list</td>
<td>File containing the names of specific regions to analyze, these target names must match the names in the bed file defining the target regions to analyze.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td>-f, --filter_list</td>
<td>A file containing specific variant events to filter out.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td>-n, --nprocessors</td>
<td>The number of processors to use for analysis.</td>
<td>1</td>
</tr>
<tr>
<td></td>
<td>-s, --start_blat_sever</td>
<td>Option to indicate that a blat server needs to be started before analysis begins. A random part number and the localhost will be used if neither is specified.</td>
<td>False</td>
</tr>
<tr>
<td></td>
<td>-k, --keep_blat_server</td>
<td>Option to keep the blat server running after the analysis completes.</td>
<td>False</td>
</tr>
<tr>
<td></td>
<td>-p, --port_number</td>
<td>The port number for the blat server to either start on or already is running on.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td>-c, --config</td>
<td>The configuration filename that contains additional parameters.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>start_blat_server</td>
<td></td>
<td>Start the blat server prior to performing the analysis</td>
<td>NA</td>
</tr>
<tr>
<td></td>
<td>-h, --help</td>
<td>Show the help message and exit.</td>
<td>NA</td>
</tr>
<tr>
<td></td>
<td>--hostname</td>
<td>The hostname for the blat server.</td>
<td>localhost</td>
</tr>
<tr>
<td></td>
<td>-p, --port_number</td>
<td>The port number for the blat server to either start on or already is running on.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td>-c, --config</td>
<td>The configuration filename that contains additional parameters.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>prepare_reference_data</td>
<td></td>
<td>Extract the reference sequence from all the input target regions and store them in files to access during analysis.</td>
<td>NA</td>
</tr>
<tr>
<td></td>
<td>-h, --help</td>
<td>Show the help message and exit</td>
<td>NA</td>
</tr>
<tr>
<td></td>
<td>-g, --gene_list</td>
<td>File containing the names of specific regions to analyze, these target names must match the names in the bed file defining the target regions to analyze.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td>-c, --config</td>
<td>The configuration filename that contains additional parameters.</td>
<td>None</td>
</tr>
<tr>
<td></td>
<td>-n, --nprocessors</td>
<td>The number of processors to use for analysis.</td>
<td>1</td>
</tr>
</tbody>
</table>

<h2>
<a id="output-files-and-formats" class="anchor" href="#output-files-and-formats" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output files and formats</h2>

<ul>
<li>While the program is running a log file analysis_dir/log.txt will continually be updated with information regarding the status of the analysis. </li>
</ul>

<h3>
<a id="logging" class="anchor" href="#logging" aria-hidden="true"><span class="octicon octicon-link"></span></a>Logging</h3>

<ul>
<li>While the program is running a log file will (&lt;analysis_directory&gt;/log.txt) continually be updated with information regarding the status of the analysis.</li>
</ul>

<h3>
<a id="target-data" class="anchor" href="#target-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Target data</h3>

<ul>
<li>For each region analyzed, a directory with the name of the region, as specified in the targets bed file, is created in a 'targets' directory (&lt;analysis_dir&gt;/targets).</li>
<li>Each target directory contains 'data', 'contigs', and 'kmers' directories. 

<ul>
<li>data - Contains the extracted reference-aligned reads from this target region as well as the kmers created from these extracted reads.</li>
<li>contigs - Contains a directory for each contig that was created, which contains the contig sequence in fasta format, and the reads used to assemble the contig in fastq format. The BLAT results are also stored with formatted output if a SV was called.</li>
<li>kmers - Contains the file with all the sample-only kmer sequences and how many reads in which it was contained as well as a file with the kmers and read ids that were used in assembling each contig.</li>
</ul>
</li>
</ul>

<h3>
<a id="final-output" class="anchor" href="#final-output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Final output</h3>

<ul>
<li>When the program completes, the final output files are directed into a directory labeled 'output' within the specified analysis directory. </li>
<li>Summary file

<ul>
<li>A summary file labeled &lt;analysis_name&gt;_summary.out contains columns: Target name, number of contigs assembled, total number of variants detected, number of indels, number of inversions and tandem duplications, number of translocations, and a list of translocation gene partners.</li>
</ul>
</li>
<li>Output for each SV type are put in respective tab-delimited files, labeled &lt;analysis_name&gt;_&lt;indel,trl,inv_rearrangement,td_rearrangement&gt;_svs.out

<ul>
<li>The columns are:

<ul>
<li>Target_Name - Target region name</li>
<li>SV_type - Rearrangement type (rearrangement, indel)</li>
<li>SV_subtype - Rearrangement subtype (trl, inversion, tandem duplication)</li>
<li>Description - Description of the indel (I5 = Insertion of 5 base pairs, D5 = deletion of 5 base pairs)</li>
<li>All_genomic_breakpoints - All the genomic breakpoints for the structural variant event.</li>
<li>Target_genomic_breakpoints - Breakpoints only within the target region.</li>
<li>Split_read_counts - Number of assembled reads that cover the assembled contig at the inferred breakpoint.</li>
<li>Discordant_read_counts - Number of discordantly-mapped paired-end reads that support the event (Not applicable to indels).</li>
<li>Read_depth_at_genomic_breakpoints - Depth of coverage at the inferred breakpoints.</li>
<li>Align_cigar - Cigar formatted string to indicate realignment to the reference sequence.</li>
<li>Strands - Strands the contig sequence realigned to on the reference sequence.</li>
<li>Total_mismatches - Number of base pairs that were mismatched in the realignment to the reference sequence.</li>
<li>Realignment_uniqueness - Uniqueness metric for the realignment.</li>
<li>Contig_ID - Contig ID</li>
<li>Contig_length - Number of base pairs in the contig sequence.</li>
<li>Contig_sequence - Contig sequence.</li>
<li>Filtered - Boolean field to indicate if the variant passed basic filtering criteria.</li>
<li>Filtered_reason - Message for filter reason if Filtered is True.</li>
<li>Filter_values - Additional information from the realignment.</li>
</ul>
</li>
</ul>
</li>
<li>Each target gene in which a SV was detected has a separate output directory (&lt;analysis_dir&gt;/output/&lt;target_name&gt;) containing formatted output specific to the target and the related reference-aligned sequence reads for the contigs that contain the structural variants detected in BAM format.</li>
</ul>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/a-bioinformatician">a-bioinformatician</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-67035322-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
